warehouse/oidc/views.py

```diff
@@ -88,7 +70,85 @@ def _invalid(errors):
+    pending_provider = oidc_service.find_provider(claims, pending=True)
+    if pending_provider is not None:
+        factory = ProjectFactory(request)
+
+        # If the project already exists, this pending provider is no longer
+        # valid and needs to be removed.
+        # NOTE: This is mostly a sanity check, since pending providers should
+        # be disposed of as part of the ordinary project creation flow.
+        if pending_provider.project_name in factory:
+            request.db.delete(pending_provider)
+            return _invalid(
+                errors=[
+                    {
+                        "code": "invalid-pending-provider",
+                        "description": "valid token, but project already exists",
+                    }
+                ]
+            )
+
+        # If the project doesn't exist, then we need to:
+        # 1. Create it (with the pending provider's user as an owner);
+        # 2. Reify the pending provider into a normal OIDC provider;
+        # 3. Add the reified provider to the newly created project.
+
+        # TODO: Dedupe this with `utils.project.add_project`?
+        new_project = Project(name=pending_provider.project_name)
+        request.db.add(new_project)
+
+        request.db.add(
+            JournalEntry(
+                name=new_project.name,
+                action="create",
+                submitted_by=pending_provider.added_by,
+                submitted_from=request.remote_addr,
+            )
+        )
+        new_project.record_event(
+            tag=EventTag.Project.ProjectCreate,
+            ip_address=request.remote_addr,
+            additional={"created_by": pending_provider.added_by.username},
+        )
+
+        request.db.add(
+            Role(user=pending_provider.added_by, project=new_project, role_name="Owner")
+        )
+
+        # TODO: This should be handled by some sort of database trigger or a
+        #       SQLAlchemy hook or the like instead of doing it inline in this
+        #       view.
+        request.db.add(
+            JournalEntry(
+                name=new_project.name,
+                action=f"add Owner {pending_provider.added_by.username}",
+                submitted_by=pending_provider.added_by,
+                submitted_from=request.remote_addr,
+            )
+        )
+        new_project.record_event(
+            tag=EventTag.Project.RoleAdd,
+            ip_address=request.remote_addr,
+            additional={
+                "submitted_by": pending_provider.added_by.username,
+                "role_name": "Owner",
+                "target_user": pending_provider.added_by.username,
+            },
+        )
+
+        new_provider = pending_provider.reify()
+        new_project.oidc_providers.append(new_provider)
+
+        # From here, the pending OIDC provider is vestigial, so we can remove it.
+        request.db.delete(pending_provider)
+        request.db.flush()
```

Comment:
Flagging: This is partially duplicated with `add_project` and the upload endpoint, as well as with the normal provider creation logic. But there isn't really a great place to dedupe it against any of those, since each makes various contextual assumptions (e.g. that the `request.identity` is a user).

Reply:
TODO: Factor this into a service method.

Reply:
[6ec7f22](https://github.com/pypi/warehouse/pull/12915/commits/6ec7f22787e0123a4993219e94b3853d7e1ac0d3)